js/modules/k6/ws/ws_test.go

```diff
@@ -504,3 +539,53 @@ func TestTLSConfig(t *testing.T) {
 	})
 	assertSessionMetricsEmitted(t, stats.GetBufferedSamples(samples), "", sr("WSSBIN_URL/ws-close"), 101, "")
 }
+
+func TestReadPump(t *testing.T) {
+	var closeCode int
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		conn, err := (&websocket.Upgrader{}).Upgrade(w, r, w.Header())
+		assert.NoError(t, err)
+		closeMsg := websocket.FormatCloseMessage(closeCode, "")
+		_ = conn.WriteControl(websocket.CloseMessage, closeMsg, time.Now().Add(time.Second))
+	}))
+	defer srv.Close()
+
+	closeCodes := []int{websocket.CloseNormalClosure, websocket.CloseGoingAway, websocket.CloseInternalServerErr}
+
+	numAsserts := 0
+	srvURL := "ws://" + srv.Listener.Addr().String()
+
+	// Ensure readPump returns the response close code sent by the server
+	for _, code := range closeCodes {
+		closeCode = code
+		conn, resp, err := websocket.DefaultDialer.Dial(srvURL, nil)
+		assert.NoError(t, err)
+		defer func() {
+			_ = resp.Body.Close()
+			_ = conn.Close()
+		}()
+
+		msgChan := make(chan []byte)
+		errChan := make(chan error)
+		closeChan := make(chan int)
+		go readPump(conn, msgChan, errChan, closeChan)
+
+	readChans:
+		for {
+			select {
+			case responseCode := <-closeChan:
+				assert.Equal(t, code, responseCode)
+				numAsserts++
+				break readChans
+			case <-errChan:
+				continue
+			case <-time.After(time.Second):
+				t.Errorf("Read timed out")
+				break readChans
+			}
+		}
+	}
+
+	// Ensure all close code asserts passed
+	assert.Equal(t, numAsserts, len(closeCodes))
```

Comment:
This is probably good idea to be with [subTests](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks) it will mean that that defer will execute between the different closecodes tests ;).

Reply:
Thanks, resolved in ded5fe3b.